src/main/java/graphql/analysis/QueryTraversal.java

```diff
@@ -47,29 +48,43 @@
     private final ValuesResolver valuesResolver = new ValuesResolver();
     private final SchemaUtil schemaUtil = new SchemaUtil();
     private final ChildrenOfSelectionProvider childrenOfSelectionProvider;
+    private final GraphQLObjectType rootParentType;
 
     public QueryTraversal(GraphQLSchema schema,
```

Comment:
Put a Builder on this and make the constructor private.  This allows for better API evolution over time

Reply:
mh ... I get your point but the arguments are not sometimes mutually exclusive: for example `operation` is only valid with a `Document`, same with `root` and `rooParentType` etc.

Is it worth extracting a whole `QueryTraversalBuilder`? Other options?

Reply:
I think generally the graphql-java should use builders on PublicAPI elements because without it the constructor becomes public API and over loading is your only way to construct things

In this case I would do `QueryTraversal.newQueryTraversal()` and have an internal builder of them